feat(upgrade): support bindToController with binding definitions

0x-r4bbit opened the pull request:
Since angular `1.4` we can also pass controller bindings directly to `bindToController`, making this syntax more convenient

This needs to be implemented I just added a test for it now.

/cc @mhevery

mhevery commented:
@PascalPrecht good catch, could you give the implementation a try?

PascalPrecht commented:
@mhevery I can try but surely need some guidance. Oh and earliest after AngularConnect :)

0x-r4bbit committed:
feat(upgrade): support bindToController with binding definitions

Since angular 1.4 we can also pass controller bindings directly to bindToController, making this syntax more convenient

Closes #4784

PascalPrecht commented:
@mhevery this is ready for review.

/cc @IgorMinar

PascalPrecht commented:
I realised that my implementation takes either `scope` bindings or `bindToController` bindings. However, technically, in an Angular 1 component, you could use both. E.g.

```
app.directive('foo', function () {
  return {
    bindToController: {
      bar: '@'
    },
    scope: {
      barFoo: '@'
    },
    controller: 'FooController as foo',
    template: '{{foo.bar}} {{barFoo}}'
  };
});
```

As far as I understand, `UpgradeAdapter` currently only supports a single destination object where values of bindings are written to (either controller or scope).

How do we want to deal with that? What do we want to support?

mhevery commented:
@PascalPrecht I don't think `ngUpgrade` will ever support / mimic the Angular 1 exactly. (For example we don't and can't support ng1 replace feature) So in this case I think we should just check for this corner case and if this is what the user has requested throw a descriptive error.

alxhub commented:
Hi @PascalPrecht,

Please fix Travis errors and then I can merge this PR.

Thanks!

PascalPrecht commented:
@alxhub Alrighty.

So should I not add an error case as @mhevery suggested?

alxhub commented:
@PascalPrecht: Please do address @mhevery's comments - it just looks like the "pr_action: merge" label was applied prematurely. That's a signal in our process that a PR is ready to be merged.

PascalPrecht commented:
@alxhub this is implemented now. Waiting for travis to be happy.

PascalPrecht commented:
Yay, travis is happy. Let's merge it away.

teropa commented:
+1 - since `module.component` in 1.5 always uses `bindToController`, this is needed for any bindings in such components to work.

PascalPrecht commented:
Oh i thought this has been merged already.

choeller commented:
+1 The 1.5 component-API is also not working with upgradeNg1Component currently:
```
 .component('checked', {
    bindings: {
      check: '='
    },
    template: 'Is checked: {{ctrl.check}}',
    controller: function() {}
  });
```
in combination with
```
      <checked [check]="isChecked"></checked>
```
currently leads to:
```
Can't bind to 'check' since it isn't a known native property ("
      <checked [ERROR ->][check]="isChecked"></checked>
```

So currently also the docs on https://angular.io/docs/ts/latest/guide/upgrade.html are saying above is working

mattdsteele commented:
Until this gets merged, is there a good way to get `upgradeNg1Component` to function with bindings in Angular 1 code? Based on [this code in the Playground module](https://github.com/angular/angular/blob/master/modules/playground/src/upgrade/index.ts#L27-L35), you should be able to define the Angular 1 directive's bindings on an isolate scope, but I can't get these to work with the UpgradeAdapter, either.

/cc @choeller @teropa

PascalPrecht commented:
@mattdsteele This should work fine as it's also covered in unit tests. Can you create a plunk demonstrate what doesn't work?

mattdsteele commented:
@PascalPrecht Seems like it's working once I upgraded to beta.3 - so no worries. Thanks for merging!
